### ISSUE TITLE
Set-AzureRmVMExtension throws 'Changing Property publisher is not allowed'

### ISSUE THREAD

**Issue by travismc1** [Azure/azure-powershell#4130]
### Cmdlet(s)
Following the powershell for vm access reset: https://github.com/Azure/azure-linux-extensions/blob/master/VMAccess/README.md

$RGName = '<resource-group-name>'
$VmName = '<vm-name>'
$Location = '<location>'

$ExtensionName = 'VMAccessForLinux'
$Publisher = 'Microsoft.OSTCExtensions'
$Version = '<version>'

$PublicConf = '{}'
$PrivateConf = '{
  "username": "<username>",
  "password": "<password>",
  "ssh_key": "<cert-contents>",
  "reset_ssh": true|false,
  "remove_user": "<username-to-remove>"
}'

Set-AzureRmVMExtension -ResourceGroupName $RGName -VMName $VmName -Location $Location `
  -Name $ExtensionName -Publisher $Publisher -ExtensionType $ExtensionName `
  -TypeHandlerVersion $Version -Settingstring $PublicConf -ProtectedSettingString $PrivateConf

### PowerShell Version
Instructions: to get PowerShell version, type `$PSVersionTable` and look for the value associated with `PSVersion`
PSVersion: 5.1.15063.296

### Module Version
3.1.0 AzureRM.Compute module

### OS Version
Instructions: to get OS version, type `$PSversionTable` and look for value associated with `BuildVersion`
10.0.15063.296

### Description
Unable to use Set-AzureRmVMExtension as it throws exception described above. It requires specifying publisher but then throws when publisher is provided.

### Debug Output
Instructions: to get Debug Output, set `$DebugPreference="Continue"` and then execute the cmdlet or script causing the issue

DEBUG: ============================ HTTP REQUEST ============================

HTTP Method:
PUT

Absolute Uri:
https://management.azure.com/subscriptions/b835f458-560c-49c2-8781-4b13d892ddd9/resourceGroups/travis613rg/providers/Microsoft.Compute/virtualMachines/travis613test/extensions/VMAccessForLinux?api-version=2016-04-30-preview

Headers:
x-ms-client-request-id        : 0b0ffe08-0777-4ca1-8959-32a87190704b
accept-language               : en-US

Body:
{
  "properties": {
    "publisher": "Microsoft.OSTCExtensions",
    "type": "VMAccessForLinux",
    "typeHandlerVersion": "1.4",
    "autoUpgradeMinorVersion": true,
    "settings": {},
    "protectedSettings": {
      "username": "travismc",
      "password": "**test-password"
    }
  },
  "location": "North Europe"
}

DEBUG: ============================ HTTP RESPONSE ============================

Status Code:
Conflict

Headers:
Pragma                        : no-cache
Strict-Transport-Security     : max-age=31536000; includeSubDomains
x-ms-served-by                : c5c25c01-22c3-45f1-a24e-1f8a30c658a2_131415255904782130
x-ms-request-id               : b0ed6f43-2967-4087-a814-d54da751c9bc
Cache-Control                 : no-cache
Server                        : Microsoft-HTTPAPI/2.0,Microsoft-HTTPAPI/2.0
x-ms-ratelimit-remaining-subscription-writes: 1199
x-ms-correlation-request-id   : c285d8f6-8cb1-4ed7-bfe4-edd643cfdca8
x-ms-routing-request-id       : WESTUS:20170613T235451Z:c285d8f6-8cb1-4ed7-bfe4-edd643cfdca8
Date                          : Tue, 13 Jun 2017 23:54:50 GMT

Body:
{
  "error": {
    "code": "PropertyChangeNotAllowed",
    "target": "publisher",
    "message": "Changing property 'publisher' is not allowed."
  }
}

DEBUG: 4:54:51 PM - SetAzureVMExtensionCommand end processing.
DEBUG: 4:54:51 PM - SetAzureVMExtensionCommand end processing.
Set-AzureRmVMExtension : Changing property 'publisher' is not allowed.
ErrorCode: PropertyChangeNotAllowed
ErrorMessage: Changing property 'publisher' is not allowed.
StatusCode: 409
ReasonPhrase: Conflict
OperationID : b0ed6f43-2967-4087-a814-d54da751c9bc
At line:1 char:1
+ Set-AzureRmVMExtension -ResourceGroupName $RGName -VMName $VmName -Lo ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : CloseError: (:) [Set-AzureRmVMExtension], ComputeCloudException
    + FullyQualifiedErrorId : Microsoft.Azure.Commands.Compute.SetAzureVMExtensionCommand

DEBUG: 4:54:51 PM - SetAzureVMExtensionCommand end processing.
DEBUG: 4:54:51 PM - SetAzureVMExtensionCommand end processing.
### Script/Steps for Reproduction
Follow link described above

**Comment by cormacpayne** [comment]
@travismc1 I have contacted the Extensions team to take a look at this issue

**Comment by ivanthelad** [comment]
Any update on this issue?

**Comment by bramvdklinkenberg** [comment]
Same issue here. I followed the instructions on [https://docs.microsoft.com/en-us/azure/virtual-machines/linux/diagnostic-extension](url) and get the same error.

**Comment by maddieclayton** [comment]
@danielsollondon Any update on this issue?

**Comment by bramvdklinkenberg** [comment]
Got it to work with a shell script (I know, not powershell.. ;) )
But wanted to share any way how to do it with the azure cli.

`#!/bin/bash

RG="$1"
STORAGE_ACCOUNT="$2"
VM=$(az vm list --resource-group ${RG} | grep node | awk '{ print $1 }')

for i in ${VM}; do
    VM_ID=$(az vm show -g ${RG} -n $i --query "id" -o tsv)
    wget https://raw.githubusercontent.com/Azure/azure-linux-extensions/master/Diagnostic/tests/lad_2_3_compatible_portal_pub_settings.json -O portal_public_settings.json
    sed -i "s#__DIAGNOSTIC_STORAGE_ACCOUNT__#${VM_ID}#g" portal_public_settings.json
    sed -i "s#__VM_RESOURCE_ID__#${VM_ID}#g" portal_public_settings.json

    STORAGE_SASTOKEN=$(az storage account generate-sas --account-name ${STORAGE_ACCOUN} --expiry 9999-12-31T23:59Z --permissions wlacu --resource-types co --services bt -o tsv)
    PROTECTED_SETTINGS="{'storageAccountName': '${STORAGE_ACCOUNT}', 'storageAccountSasToken': '${STORAGE_SASTOKEN}'}"

    az vm extension set --publisher Microsoft.Azure.Diagnostics --name LinuxDiagnostic --version 3.0 --resource-group ${RG} --vm-name $i --protected-settings "${PROTECTED_SETTINGS}" --settings portal_public_settings.json; done
done

exit 0`

**Comment by ankitmundra88** [comment]
I am also facing the same issue

From within an Ubuntu 16.04
Azure CLI version 2.0.32

**Comment by bramvdklinkenberg** [comment]
@ankitmundra88 , make sure you use "--publisher Microsoft.Azure.Diagnostics"

**Comment by ankitmundra88** [comment]
@bramvdklinkenberg Thanks for the response. I managed to get it working.

Below is the observation ::

- I already had the Linux Diagnostics Extension Enabled on the VM. That extension was from the provider Microsoft.OSTCExtensions and the extension version 2.*. 
- After uninstallation of the previous extension I was able to successfully install the 'LinuxDiagnostic' version 3.0.0 extension.

I think the issue was it was trying to upgrade the previous LinuxDiagnostic extenstion which is from a different publisher.
